### ISSUE TITLE
what is meaning of    self.query_embedding = nn.Embedding(self.num_query, self.embed_dims * 2)

### ISSUE THREAD

**Issue by hitbuyi** [WangYueFt/detr3d#72]
in detr3d_head.py,   
```
       if not self.as_two_stage:
            self.query_embedding = nn.Embedding(self.num_query,
                                                self.embed_dims * 2)

```

why  self.embed_dims  multiplfies *2?  what's the meaning here?

**Comment by Raiden-cn** [comment]
I guess it initializes object query like DETR.

**Comment by icefrogplus** [comment]
half 256 is _query_pos_ to predict the initial (x0, y0, z0) ; 
where the code in  _detr3d_transformer.py_

`    

def init_layers(self):   

        """Initialize layers of the Detr3DTransformer."""  

        self.reference_points = nn.Linear(self.embed_dims, 3)    

`

and we split the 512 to the 2*256:

`
        query_pos, query = torch.split(query_embed, self.embed_dims , dim=1)  

`

so the _query_pos_  to predict an initial 3D position 

`

        # 生成初始 3D 参考点  

        reference_points = self.reference_points(query_pos)     # [bs, 900, 256] -> [bs, 900, 3]  

        reference_points = reference_points.sigmoid()           # 0~1 归一化位置坐标（x, y, z）  

        init_reference_out = reference_points                   # 初始化 3d 点坐标 [bs, num_query, 3]  

`

left the _query_ is the common query to Transformer Attention operation, also in the  _detr3d_transformer.py_

`

@ATTENTION.register_module()  

class Detr3DCrossAtten(BaseModule):  

`

which get 

`

        # 输出残差融合: 跨视角注意力融合的特征 + 残差连接（原 query）+ 位置编码特征  

        return self.dropout(output) + inp_residual + pos_feat  # [num_query, bs, embed_dims] = [900, 1, 256]  

`

then, feed into a _reg_branches_ to predict (δx, δy, δz), to iterative update the prediction (x + δx, y + δy, z + δz)

**Comment by Raiden-cn** [comment]
It make sense.

**Comment by linzangsc** [comment]
> half 256 is _query_pos_ to predict the initial (x0, y0, z0) ; where the code in _detr3d_transformer.py_
> 
> `
> 
> def init_layers(self):
> 
> ```
>     """Initialize layers of the Detr3DTransformer."""  
> 
>     self.reference_points = nn.Linear(self.embed_dims, 3)    
> ```
> 
> `
> 
> and we split the 512 to the 2*256:
> 
> ` query_pos, query = torch.split(query_embed, self.embed_dims , dim=1)
> 
> `
> 
> so the _query_pos_ to predict an initial 3D position
> 
> `
> 
> ```
>     # 生成初始 3D 参考点  
> 
>     reference_points = self.reference_points(query_pos)     # [bs, 900, 256] -> [bs, 900, 3]  
> 
>     reference_points = reference_points.sigmoid()           # 0~1 归一化位置坐标（x, y, z）  
> 
>     init_reference_out = reference_points                   # 初始化 3d 点坐标 [bs, num_query, 3]  
> ```
> 
> `
> 
> left the _query_ is the common query to Transformer Attention operation, also in the _detr3d_transformer.py_
> 
> `
> 
> @ATTENTION.register_module()
> 
> class Detr3DCrossAtten(BaseModule):
> 
> `
> 
> which get
> 
> `
> 
> ```
>     # 输出残差融合: 跨视角注意力融合的特征 + 残差连接（原 query）+ 位置编码特征  
> 
>     return self.dropout(output) + inp_residual + pos_feat  # [num_query, bs, embed_dims] = [900, 1, 256]  
> ```
> 
> `
> 
> then, feed into a _reg_branches_ to predict (δx, δy, δz), to iterative update the prediction (x + δx, y + δy, z + δz)

however in detr, object query is initialized to zero for each batch. why 

> half 256 is _query_pos_ to predict the initial (x0, y0, z0) ; where the code in _detr3d_transformer.py_
> 
> `
> 
> def init_layers(self):
> 
> ```
>     """Initialize layers of the Detr3DTransformer."""  
> 
>     self.reference_points = nn.Linear(self.embed_dims, 3)    
> ```
> 
> `
> 
> and we split the 512 to the 2*256:
> 
> ` query_pos, query = torch.split(query_embed, self.embed_dims , dim=1)
> 
> `
> 
> so the _query_pos_ to predict an initial 3D position
> 
> `
> 
> ```
>     # 生成初始 3D 参考点  
> 
>     reference_points = self.reference_points(query_pos)     # [bs, 900, 256] -> [bs, 900, 3]  
> 
>     reference_points = reference_points.sigmoid()           # 0~1 归一化位置坐标（x, y, z）  
> 
>     init_reference_out = reference_points                   # 初始化 3d 点坐标 [bs, num_query, 3]  
> ```
> 
> `
> 
> left the _query_ is the common query to Transformer Attention operation, also in the _detr3d_transformer.py_
> 
> `
> 
> @ATTENTION.register_module()
> 
> class Detr3DCrossAtten(BaseModule):
> 
> `
> 
> which get
> 
> `
> 
> ```
>     # 输出残差融合: 跨视角注意力融合的特征 + 残差连接（原 query）+ 位置编码特征  
> 
>     return self.dropout(output) + inp_residual + pos_feat  # [num_query, bs, embed_dims] = [900, 1, 256]  
> ```
> 
> `
> 
> then, feed into a _reg_branches_ to predict (δx, δy, δz), to iterative update the prediction (x + δx, y + δy, z + δz)

Object query is initialized to zero for each bach in DETR. However here in DETR3D, object query becomes learnable embedding. Why and what's the difference?

**Comment by icefrogplus** [comment]
> > half 256 is _query_pos_ to predict the initial (x0, y0, z0) ; where the code in _detr3d_transformer.py_半 256 _query_pos_ 预测初始（x0， y0， z0）;其中代码_detr3d_transformer.py_
> > `
> > def init_layers(self):  自 init_layers（自尊）：
> > ```
> >     """Initialize layers of the Detr3DTransformer."""  
> > 
> >     self.reference_points = nn.Linear(self.embed_dims, 3)    
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `
> > and we split the 512 to the 2*256:我们将512分成2×256：
> > ` query_pos, query = torch.split(query_embed, self.embed_dims , dim=1)' query_pos， query = torch.split（query_embed， self.embed_dims， dim=1）
> > `
> > so the _query_pos_ to predict an initial 3D position因此，预测初始三维位置 _query_pos_
> > `
> > ```
> >     # 生成初始 3D 参考点  
> > 
> >     reference_points = self.reference_points(query_pos)     # [bs, 900, 256] -> [bs, 900, 3]  
> > 
> >     reference_points = reference_points.sigmoid()           # 0~1 归一化位置坐标（x, y, z）  
> > 
> >     init_reference_out = reference_points                   # 初始化 3d 点坐标 [bs, num_query, 3]  
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `
> > left the _query_ is the common query to Transformer Attention operation, also in the _detr3d_transformer.py_左_查询_是变换器注意力作的常见查询，也包含在_detr3d_transformer.py_
> > `
> > @ATTENTION.register_module()
> > class Detr3DCrossAtten(BaseModule):class Detr3DCrossAtten（BaseModule）：
> > `
> > which get  该得
> > `
> > ```
> >     # 输出残差融合: 跨视角注意力融合的特征 + 残差连接（原 query）+ 位置编码特征  
> > 
> >     return self.dropout(output) + inp_residual + pos_feat  # [num_query, bs, embed_dims] = [900, 1, 256]  
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `
> > then, feed into a _reg_branches_ to predict (δx, δy, δz), to iterative update the prediction (x + δx, y + δy, z + δz)然后，输入要预测的 _reg_branches_（δx，δy，δz），迭代更新预测（x+δx，y+δy，z+δz）
> 
> however in detr, object query is initialized to zero for each batch. why但在 Detr 中，每个批次的对象查询初始化为零。为什么
> 
> > half 256 is _query_pos_ to predict the initial (x0, y0, z0) ; where the code in _detr3d_transformer.py_半 256 _query_pos_ 预测初始（x0， y0， z0）;其中代码_detr3d_transformer.py_
> > `
> > def init_layers(self):  自 init_layers（自尊）：
> > ```
> >     """Initialize layers of the Detr3DTransformer."""  
> > 
> >     self.reference_points = nn.Linear(self.embed_dims, 3)    
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `
> > and we split the 512 to the 2*256:我们将512分成2×256：
> > ` query_pos, query = torch.split(query_embed, self.embed_dims , dim=1)' query_pos， query = torch.split（query_embed， self.embed_dims， dim=1）
> > `
> > so the _query_pos_ to predict an initial 3D position因此，预测初始三维位置 _query_pos_
> > `
> > ```
> >     # 生成初始 3D 参考点  
> > 
> >     reference_points = self.reference_points(query_pos)     # [bs, 900, 256] -> [bs, 900, 3]  
> > 
> >     reference_points = reference_points.sigmoid()           # 0~1 归一化位置坐标（x, y, z）  
> > 
> >     init_reference_out = reference_points                   # 初始化 3d 点坐标 [bs, num_query, 3]  
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `
> > left the _query_ is the common query to Transformer Attention operation, also in the _detr3d_transformer.py_左_查询_是变换器注意力作的常见查询，也包含在_detr3d_transformer.py_
> > `
> > @ATTENTION.register_module()
> > class Detr3DCrossAtten(BaseModule):class Detr3DCrossAtten（BaseModule）：
> > `
> > which get  该得
> > `
> > ```
> >     # 输出残差融合: 跨视角注意力融合的特征 + 残差连接（原 query）+ 位置编码特征  
> > 
> >     return self.dropout(output) + inp_residual + pos_feat  # [num_query, bs, embed_dims] = [900, 1, 256]  
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `
> > then, feed into a _reg_branches_ to predict (δx, δy, δz), to iterative update the prediction (x + δx, y + δy, z + δz)然后，输入要预测的 _reg_branches_（δx，δy，δz），迭代更新预测（x+δx，y+δy，z+δz）
> 
> Object query is initialized to zero for each bach in DETR. However here in DETR3D, object query becomes learnable embedding. Why and what's the difference?在 DETR 中，每个 bach 的对象查询初始化为零。但在 DETR3D 中，对象查询变成了可学习的嵌入。为什么？有什么区别？

In DETR3D, we init 900 query with the shape [900, 256], forward with a nn.linear(256, 3) to init the 3D box center points (x0, y0, z0) , each epoch or iteration training backward to amend the init 3D points center (x0, y0, z0), the object query maybe the same?
